### PR TITLE
Update all MCUs to ChibiOS 21.11.x/master 

### PR DESCRIPTION
**Pull request by zykrah**
This PR updates ChibiOS-Contrib to be up to date with ChibiOS 21.11.x (the master branch, because there is no separate branch for 21.11.x currently). Additionally, I believe this is the first build to include the new Raspberry Pi Pico RP2040 MCU.

This is in response to https://github.com/qmk/qmk_firmware/pull/14877. I assume by merging this PR, development on getting RP2040 officially supported by QMK (without having to use sekigon_gonnoc's RP2040 VIAL fork) will be allowed to continue

From @tzarc in the QMK discord:
image
image

Changes:

- Modified `build.yml` to use master branch of ChibiOS. Also fixed a minor bug that didn't allow extension packages to export.
- Updated `chconf.h` files of MCUs, adding new required definintions going from ChibiOS 20.3.x -> ChibiOS 21.11.x 
- Some `halconf.h` files were also updated
- Updated some `mcuconf.h` files (mainly STM32F4xx)
- Updated missing MCU definitions from some files (mainly some STM `mcuconf.h` files)
- Updated some STM `ffconf.h` files
- Updated some old/deprecated method names
- Updated some other minor code deprecations
- I have removed `PACKED_VAR` from some typedef structs in a file because of redefinition errors. I don't know if this is the correct way of resolving this.

Notes: 

- **I actually do not have much experience with C**. Pretty much all of this was trial and error, so I may have done some things wrong.
- These changes will have to be merged into a new branch, ie `chibios-21.11.x`
- MCUs may all need to be individually tested in real life to see if everything functions properly. All support is currently theoretical.
- There are a lot of commits because I still don't actually know how to build locally. (A heads up on how to do this would be nice)
- I chose the master branch of ChibiOS because no branch for 21.11.x (stable) yet exists (This will have to change in the future)
- ~~I'll have to remove some of those vscode files.~~
- Some STM32F4XX mcuconf.h files may still need slight updating + checking (check against the ChibiOS demos)
- I'm unsure if some of the changes made to some `chconf.h` files using ChibiOS NIL are necessary.
- I did not go over any of the warnings/notes presented when building CI, i only fixed any errors that prevented CI from building. Hence, there may still be some glaring issues that may have to be fixed additionally.
- RP2040 is still untested afaik. @hanya may have already tested their code before it was merged though.
- After making this PR, I have looked and many of these changes mirror other PRs regarding updating MCUs to 21.6.x (However this PR seeks to update to master branch/21.11.x rather than just 21.6.x)
- I forgot to add myself to AUTHORS.txt
- Thank you to all who helped me out, both on the ChibiOS discord and the AusKeys discord server

### CODE REVIEW

**Comment by zykrah** [comment]
Note: From what I can see the LGTM analysis is failing because it is still using ChibiOS 20.3.x. I don't know how to modify this.

**Comment by zykrah** [comment]
https://github.com/ChibiOS/ChibiOS-Contrib/pull/305 Better explains some of the changes I have made:

- Changes the fault handler to gather info from the current core (the "ch" error i reference in the commits)
- Updates the `Makefile` to include the `ARMv7-M` port which replaces the generic `ARMCMx` port (The same was done for any code that uses the `ARMv6-M` port)
- Updates the `chconf.h` to define newly introduced settings

File: .github/workflows/build.yml
Comment (zykrah) [inline .github/workflows/build.yml]: This is temporary. It will have to be changed to `stable_21.11.x` when the branch is available.

image

File: .github/workflows/build.yml
Comment (zykrah) [inline .github/workflows/build.yml]: May have to be changed to just `chibios-21.11.x`. I don't know what to do with this one.

File: .github/workflows/build.yml
Comment (zykrah) [inline .github/workflows/build.yml:28]: This is to fix a problem with a whitespace existing in one of the files that gets extracted (pico-sdk).

File: os/common/ports/ARMCMx/compilers/GCC/utils/fault_handlers_v7m.c
Comment (zykrah) [inline os/common/ports/ARMCMx/compilers/GCC/utils/fault_handlers_v7m.c:18]: Unsure if this change is correct

File: os/common/ports/ARMCMx/compilers/GCC/utils/fault_handlers_v7m.c
Comment (zykrah) [inline os/common/ports/ARMCMx/compilers/GCC/utils/fault_handlers_v7m.c:136]: `currcore->` points to the current core. Applies to other changes in this file too.

image

File: os/hal/include/hal_usb_msd.h
Comment (zykrah) [inline os/hal/include/hal_usb_msd.h:104]: Changed as a result of redefinition errors (there are more of the same changes in this file). I don't know if this is the correct way to solve this.

File: os/hal/ports/RP/LLD/USBDv1/hal_usb_lld.c
Comment (zykrah) [inline os/hal/ports/RP/LLD/USBDv1/hal_usb_lld.c:735]: As in https://github.com/ChibiOS/ChibiOS-Contrib/pull/301

File: os/various/lib_scsi.h
Comment (zykrah) [inline os/various/lib_scsi.h:113]: Changed as a result of redefinition errors (there are more of the same changes in this file). I don't know if this is the correct way to solve this.

File: testhal/RP/RP2040/RT-RP2040-PICO-HID/Client/test-usb-hid.c
Comment (zykrah) [inline testhal/RP/RP2040/RT-RP2040-PICO-HID/Client/test-usb-hid.c]: Unnecessary change. Was in an attempt to fix a warning.

File: testhal/RP/RP2040/RT-RP2040-PICO-SERIAL/Makefile
Comment (zykrah) [inline testhal/RP/RP2040/RT-RP2040-PICO-SERIAL/Makefile]: Needs to be reverted

Comment (zykrah) [reply testhal/RP/RP2040/RT-RP2040-PICO-SERIAL/Makefile]: reverted in 38c01bd

File: testhal/RP/RP2040/RT-RP2040-PICO-HID/Client/test-usb-hid.c
Comment (zykrah) [reply testhal/RP/RP2040/RT-RP2040-PICO-HID/Client/test-usb-hid.c]: reverted in 38c01bd

File: testhal/NUMICRO/NUC123/NUTINY-SDK-NUC123-V2.0/USB_HID/main.c
Comment (zykrah) [inline testhal/NUMICRO/NUC123/NUTINY-SDK-NUC123-V2.0/USB_HID/main.c:58]: Just to change to an updated method

**Comment by Snipeye** [comment]
Hello!  I've got a CS minor (and done plenty of C/C++) but I've got next to no familiarity with this library - that said, I'm a keyboard nerd and would love to see this PR completed so the RP2040 would be useable in QMK.  Is there anything that I can do to help?

**Comment by zykrah** [comment]
> Hello! I've got a CS minor (and done plenty of C/C++) but I've got next to no familiarity with this library - that said, I'm a keyboard nerd and would love to see this PR completed so the RP2040 would be useable in QMK. Is there anything that I can do to help?

Hey thanks for the expression of interest! As far as I know, all the work has already been done for the PR to be merged, however it's just waiting on a review from a maintainer. But feel free to go over the code and review it yourself if you want to check. I've already gone over all the file changes myself and everything looked good to me, apart from the few things i mentioned in the `Notes` section in the original PR message.

**Comment by rusefillc** [comment]
@fpoussin poke :)

**Comment by nodupe** [comment]
Hi, kindly, can we have some updates on this? Thank you!

**Comment by fpoussin** [comment]
Having a look

File: .github/workflows/build.yml
Comment (fpoussin) [reply .github/workflows/build.yml:19]: Master is a no go, we only support stable branches

File: .github/workflows/build.yml
Comment (fpoussin) [reply .github/workflows/build.yml:28]: LGTM

File: .github/workflows/build.yml
Comment (fpoussin) [reply .github/workflows/build.yml:7]: This needs to be reverted back for CI to work and validate this PR

Comment (fpoussin) [reply .github/workflows/build.yml]: I've pushed a `chibios-21.11.x` branch, you can use that

File: .github/workflows/build.yml
Comment (fpoussin) [reply .github/workflows/build.yml]: I've added `stable_21.11.x`, please use that, master is a no go

File: os/common/ports/ARMCMx/compilers/GCC/utils/fault_handlers_v7m.c
Comment (fpoussin) [reply os/common/ports/ARMCMx/compilers/GCC/utils/fault_handlers_v7m.c:18]: LGTM

**Review by fpoussin — CHANGES_REQUESTED** [review]

**Comment by fpoussin** [comment]
New stable branch added to ChibiOS repo, you can add the requested changes now: 
https://github.com/ChibiOS/ChibiOS/tree/stable_21.11.x

**Comment by fpoussin** [comment]
I've changed the merge branch to 21.11.x

**Comment by zykrah** [comment]
Thanks for that @fpoussin, requested changes implemented. ~~I forgot to rename my branch before pushing so CI didn't run, but I only changed the build.yml as is evident in the commit. Do you want me to still run a test CI?~~

EDIT: Nevermind I just renamed the branch now. CI is running as I type this.

EDIT 2: Looks like my lack of git experience has put this in a tricky spot. Oops. Restored the branch but the CI should still be running in the 21.11.x branch in my fork. Hope that's okay.

EDIT 3: Looks like CI passed :) 🎉

**Comment by fpoussin** [comment]
Yes that's fine, CI should pass then I'll merge.
We can do more cleanup in another PR.

**Comment by zykrah** [comment]
> Yes that's fine, CI should pass then I'll merge. We can do more cleanup in another PR.

CI passed 👍 Thanks for the review fpoussin

**Review by fpoussin — APPROVED** [review]

**Comment by fpoussin** [comment]
It's done :)

**Pull request merged**
